task_processing/plugins/mesos/execution_framework.py

```diff
@@ -239,28 +245,76 @@ def unblacklist_slave(self, agent_id, timeout):
             self.blacklisted_slaves = \
                 self.blacklisted_slaves.remove(agent_id)
 
-    def enqueue_task(self, task_config):
+    def _is_pod(self, task_config):
+        if type(task_config).__name__ == 'MesosPodConfig':
+            return True
+
+        return False
+
+    def _initialize_metadata(self, task_config):
+        # task_state and task_state_history get reset every time
+        # a task is enqueued.
         with self._lock:
-            # task_state and task_state_history get reset every time
-            # a task is enqueued.
-            self.task_metadata = self.task_metadata.set(
-                task_config.task_id,
-                TaskMetadata(
-                    task_config=task_config,
-                    task_state='TASK_INITED',
-                    task_state_history=m(TASK_INITED=time.time()),
+            if self._is_pod(task_config):
+                pod_id = task_config.task_id
+                nested_container_ids = []
+
+                # We receive individual task updates for each task in
+                # task_groups. So, we are gonna keep track of all the
+                # individual task in task_metadata.
+                for task in task_config.tasks:
+                    # task_state and task_state_history get reset every time
+                    # a task is enqueued.
+                    self.task_metadata = self.task_metadata.set(
+                        task.task_id,
+                        TaskMetadata(
+                            task_config=task,
+                            pod_id=pod_id,
+                            task_state='TASK_INITED',
+                            task_state_history=m(
+                                TASK_INITED=m(
+                                    time_stamp=time.time(),
+                                    status_update=Dict()
+                                )
+                            ),
+                        )
+                    )
+                    nested_container_ids.append(task.task_id)
+
+                self._pod_to_task_mappings = self._pod_to_task_mappings.set(
```

Comment:
do we really need another datastructure for this? I know it makes for an easier lookup in task_metadata, but i'm kind of wondering if it'd be cleaner to just have a function, since in the end these are always used to access task_metadata? it reduces the two sources of truth to one:

```
def task_for_pod_ids(pod_id):
    return [tm for tm in self.task_metadata.items() if tm.value.pod_id == pod_id]
```

Reply:
Yeah, we can certainly do it this way. But, HashMap is more efficient than doing this. It is just one extra data structure.